docs/design_2x.md

```diff
@@ -1,34 +1,130 @@
+to override functions in other shared libraries (e.g., system call wrapper
+functions from libc) by supplying identically named functions with alternative
+implementations inside the shim. Note that preloading works on dynamically
+linked calls to libc system call wrappers, but not on statically linked calls
+made from inside of libc or made using a `syscall` instruction.
+
+- seccomp: System calls that are not preloadable are intercepted using the
+kernel's seccomp facility. The shim of each managed process installs a seccomp
+filter that traps all system calls (except those made in the shim) and a handler
+function to handle the trapped system calls. This facility has a very small
+overhead because it involves running the installed filter in kernel mode, but we
+infrequently incur this overhead in practice since most system calls are
+preloadable.
+
+## Emulating System Calls
+
+System calls that are intercepted by the shim (using either preloading or
+seccomp) are emulated by Shadow. Hot-path system calls (e.g., time-related
+system calls) are handled directly in the shim by using state that is stored in
+shared memory. Other system calls are sent from the shim to Shadow via the
+control channel and handled in Shadow (the shim sends the system call number and
+argument registers). While the shim is waiting for a system call to be serviced
+by Shadow, the managed process is blocked; this allows Shadow to precisely
+control the running state of each process.
+
+Shadow emulates system calls using its simulated kernel. The simulated kernel
+(re)implements (i.e., simulates) important system functionality, including: the
+passage of time; input and output operations on file, socket, pipe, timer, and
+event descriptors; packet transmissions with respect to transport layer
+protocols such as TCP and UDP; and aspects of computer networking including
+routing, queuing, and bandwidth limits. Thus, Shadow establishes a private,
+simulated network environment that is completely isolated from the real network,
+but is internally interoperable and entirely controllable.
+
+## Managing Memory
+
+Some system calls pass dynamically allocated memory addresses (e.g., the buffer
+address in the `sendto()` system call). To handle this system call in Shadow,
+this shim sends the buffer address but not the buffer contents to Shadow. Shadow
+uses an inter-process memory access manager to directly and efficiently read and
+write the memory of each managed process without extraneous data copies or
+control messages. Briefly, the memory manager (re)maps the memory of each
+managed process into a shared memory file that is accessible by both Shadow and
+the managed process. When Shadow needs to copy data from a memory address passed
+to it by the shim, the memory manager translates the managed process's memory
+address to a shared memory address and brokers requested data copies. This
+approach minimizes the number of copies needed to transfer the buffer contents
```

Comment:
Also minimizes system calls. (There are no system calls in the "mapped" access code path)